File: src/main/java/editor/database/card/CardFormat.java
package editor.database.card;

import java.time.LocalDate;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import editor.collection.CardList;
import editor.collection.deck.CategorySpec;
import editor.collection.deck.Deck;
import editor.database.attributes.CardAttribute;
import editor.util.CollectionUtils;

/**
 * This class represents a method of formatting a card into a string.  It takes
 * a format string, which consists of arbitrary characters and format specifiers,
 * and replaces the format specifiers with the corresponding values of a card
 * in a deck.  A format specifier consists of the name of a {@link CardAttribute}
 * surrounded by braces ({}).
 *
 * @author Alec Roelke
 */
public class CardFormat
{
    /**
     * This class represents a "fake" card entry that reports static data about a card.
     * It always returns 1 for count, the card's release date for date, and an empty
     * set for categories.
     *
     * @author Alec Roelke
     */
    private static class FakeEntry implements CardList.Entry
    {
        /**
         * Card so the release date can be known.
         */
        private final Card card;

        /**
         * Create a new FakeEntry.
         *
         * @param c card for the new entry
         */
        public FakeEntry(Card c)
        {
            card = c;
        }

        @Override
        public Card card()
        {
            return card;
        }

        @Override
        public Set<CategorySpec> categories()
        {
            return Collections.emptySet();
        }

        @Override
        public int count()
        {
            return 1;
        }

        @Override
        public LocalDate dateAdded()
        {
            return card.expansion().released();
        }

    }

    /**
     * Format specifier for a card.
     */
    public final String format;

    /**
     * Create a new CardFormat.
     *
     * @param pattern format specifier for a card
     */
    public CardFormat(String pattern)
    {
        format = pattern;
    }

    /**
     * Format a card using this CardFormat's format specifier.  Deck-dependent
     * values will be statically assigned:
     * {@link CardAttribute#COUNT}: 1
     * {@link CardAttribute#CATEGORIES}: empty set
     * {@link CardAttribute#DATE_ADDED}: the card's release date
     *
     * @param card card to format
     * @return the formatted string
     */
    public String format(Card card)
    {
        return format(new FakeEntry(card));
    }

    /**
     * Format a card using this CardFormat's format specifier.  Deck-dependent
     * values will be taken from the entry provided.
     *
     * @param card entry for the card to format
     * @return the formatted string
     */
    public String format(CardList.Entry card)
    {
        String pattern = format;
        for (CardAttribute type : CardAttribute.displayableValues())
        {
            String replacement = '{' + type.toString().toLowerCase() + '}';
            pattern = pattern.replace(replacement, switch (type) {
                case MANA_COST, POWER, TOUGHNESS, LOYALTY -> String.join(Card.FACE_SEPARATOR,
                    ((List<?>)card.get(type)).stream().map(String::valueOf).collect(Collectors.toList()));
                case MANA_VALUE -> String.join(Card.FACE_SEPARATOR,
                    CollectionUtils.convertToList(card.get(type), Double.class).stream().map((n) -> {
                        if (n == n.intValue())
                            return Integer.toString(n.intValue());
                        else
                            return n.toString();
                    }).collect(Collectors.toList()));
                case COLORS, COLOR_IDENTITY -> String.join(",",
                    ((List<?>)card.get(type)).stream().map(String::valueOf).collect(Collectors.toList()));
                case CATEGORIES -> String.join(",",
                    CollectionUtils.convertToSet(card.get(type), CategorySpec.class).stream().map(CategorySpec::getName).sorted().collect(Collectors.toList()));
                case DATE_ADDED -> Deck.DATE_FORMATTER.format((LocalDate)card.get(type));
                default -> String.valueOf(card.get(type));
            });
        }
        return pattern;
    }
}

File: src/main/scala/editor/database/card/CardFormat.scala
package editor.database.card

import editor.collection.CardListEntry
import editor.collection.Categorization
import editor.collection.mutable.Deck
import editor.database.attributes.CardAttribute

import java.time.LocalDate
import scala.jdk.CollectionConverters._

/**
 * A string formatter for [[Card]].  Format strings are arbitrary text with card attribute names in
 * lower case surrounded by braces.
 * 
 * @constructor create a new formatter for a string
 * @param pattern string pattern to use for formatting a card
 * 
 * @author Alec Roelke
 */
class CardFormat(pattern: String) {
  /**
   * Generate a string based on the format's pattern from the values of the attributes of a deck entry.
   * 
   * @param card deck entry to format
   * @return the formatted string
   */
  def format(card: CardListEntry): String = {
    import CardAttribute._

    CardAttribute.displayableValues.foldLeft(pattern)((p, a) => p.replace(s"{$a}".toLowerCase, a match {
      case f @ (ManaCost | EffManaValue | TypeLine | Power | Toughness | Loyalty) => f(card).mkString(Card.FaceSeparator)
      case RealManaValue => RealManaValue(card) match {
        case v if v == v.intValue => v.intValue.toString
        case v if v != v.intValue => v.toString
      }
      case c @ (Colors | ColorIdentity) => c(card).mkString(",")
      case Categories => Categories(card).map(_.name).toSeq.sorted.mkString(",")
      case DateAdded => CardAttribute.DateAdded.format(DateAdded(card))
      case _ => a(card).toString
    }))
  }

  /**
   * Generate a string based on the format's pattern from the values of the attributes of a card. Deck-specific
   * attributes are replaced with their default values.
   * 
   * @param card card to format
   * @return the formatted string
   */
  def format(c: Card): String = format(new CardListEntry {
    override val card = c
    override val categories = Set.empty[Categorization]
    override val count = 1
    override def dateAdded = card.expansion.released
  })
}
